docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,668 @@
+variant. To this end, we introduce the operator `MatchOnly`:
+
+```tla
+MatchOnly(variant, ThenOper(_)) ==
+  \* fallback untyped implementation
+  ThenOper(variant)
+```
+
+For completeness, we give the inference rule for `MatchOnly`:
+
+```
+variant: Variant(RowCons("<TAG>", r, RowNil))
+ThenOper: r => z
+------------------------------------------------ [variant_match_only]
+MatchOnly(variant, ThenOper):
+  (Variant(RowCons("<TAG>", r, RowNil)),
+   r => z
+  ) => z
+```
+
+It looks like the solution with `MatchTag` and `MatchOnly` introduce a lot of
+boilerplate. However, this is probably the best solution that we can have,
+unless we can extend the grammar of TLA+.
+
+### 5.3. Changes in the model checker
+
+Having precise types for variants, we have two options:
+
+ 1. Keep the current encoding, that is, a variant is encoded as a super-record
+ that contains all possible fields of the member records. The type checker
+ will guarantee that we do not access the fields of the super-record that are
+ not present in the actual record type.
+
+ 1. Implement the
+ [suggestion by Shon Feder](https://github.com/informalsystems/apalache/discussions/789#discussioncomment-1592118).
+
+Although Option 2 looks nicer, we prefer keeping Option 1. The reason is that
+the current implementation introduces the minimal number of constraints
+by mashing all possible fields into a super-record. The alternative solution
+(option 2) would introduce additional constraints, when the spec requires us
+to extract an element from a set.
+
+Recall the example with a set of messages:
+
+```tla
+VARIABLES
+  (*
+    @typeAlias: MESSAGE =
+       [ tag: "1a", bal: Int ]
+     | [ tag: "2a", bal: Int, val: Int ]);
+    @type: Set(MESSAGE);
+   *)  
+  msgs
+```  
+
+Consider an existential quantifier over the variable `msgs`:
+
+```tla
+Next ==
+  \E m \in msgs:
+    P
+```
+
+In the current encoding, `m` is a super-record that contains three fields:
+`tag`, `bal`, and `val`, even if some of these fields are not required by the
+actual type of `m`. In the alternative encoding, `m` is a tuple `tup`,
+which equals to one of the following tuples, depending on the value of the
+field `m.tag`:
+
+```
+tup = IF m.tag = "1a"
+      THEN << { [ tag |-> "1a", bal |-> b ] }, {} >>
+      ELSE << {}, [ tag |-> "2a", bal |-> b, val |-> v ] }, {} >>
+```
+
+Since it is impossible to statically compute the actual type of `m`, the
+rewriting rules would have to replicate the structure of both elements of the
+tuple `tup`. This would lead to a blow-up in the number of constraints.
+
+### 5.4. Additional requirements
+
+Given the decisions in Section 5.3, we additionally require that all records in
+a variant type have compatible field types. In more detail, if a variant type
+contains two record types `[ tag: "A", value: a, ... ]` and `[ tag: "B", value:
+b, ...]`, then the types `a` and `b` must be unifiable.  In practice, this
+often implies that `a` and `b` are simply the same type.
```

Comment:
This seems really strange to me: In this proposal, we will introduce fancy, row types, which are designed to support pretty remarkable flexibility and generality. Then we limit them to only being applicable to records, then we further limit them to only applying to records which have fields of the same types. It makes me feel like there must be a simpler way... If our options are so constrained, it would seem like we don't nee such generality, but if we do opt for such a general solution, then it would seem that we shouldn't need all this additional book keeping to deal with one particular case.

Reply:
I don't see row type being fancy. It is just a natural approach when you have equality constraints instead of arbitrary subtyping constraints.

Reply:
I guess I mean "fancier than non-polymorphic variants/records". 

The main thing I wish I understood around this ADR is why non-polymorphic variants and records wouldn't work for all of our needs. I apologize if I'm just being dense, but since you say records aren't extensible, and from our discussion [above](https://github.com/informalsystems/apalache/pull/1170#discussion_r771185191) it sounds like we won't support operators that are polymorphic over record fields, and now the variants supported are even more constrained than the variants supported by standard, *non-polymorphic* variant types, I can't figure out what aspect of the row typing is actually going to be used or needed.

Reply:
I'll just add my understanding of why row types are necessary here and you guys can tell me if it makes sense.

PS: I'm mixing the paper notation with the ADR notation with the intention to clarify the relation between the two things
For what I understand about the ADR and Igor's comments, the scenario where we actually need row types is this one:

```
a = [ tag |-> "t1", x |-> 2 ] :: { tag :: String, x :: Int }
b = [ tag |-> "t2", x |-> 5, y |-> "hello" ]  :: { tag :: String, x :: Int, y :: String }

msgs = { a, b } :: Set(Variant(RowCons( "t1", Rec({ x :: Int })), RowCons("t2", Rec{ x :: Int, y :: String })))
```

And this is actually a scenario I believe we need subtyping. Suppose a general type rule for Sets

```
x_0 :: t, ... x_n :: t
------------------------------- [set]
{ x_0, ..., x_n } :: Set(t)
```

for `msgs` to be correctly typed, both `{ tag :: String, x :: Int }` and `{ tag :: String, x :: Int, y :: String }` should be subtypes of `Set(Variant(RowCons( "t1", Rec({ x :: Int })), RowCons("t2", Rec{ x :: Int, y :: String })))`

The subtyping relation could also be done with a sum type. The difference between a sum type and a variant is that a variant carries a row type variable and therefore can be extended. The extension I'm referring to here is not record extension as defined in the paper, but actually the type extension that has to happen when a record of a different row type is added to a set. Therefore, in my understading, we need row types because:
1. A set can't have values of different types, but through a variant, it can have values of different row types.
2. Unification rules for sum types would not be enough to accumulate information about the record type throughout extensions such as adding a new record to a set. We need more "dynamic" unification rules such as the ones provided in the row types paper.

Reply:
Thanks, for your analysis here @gabrielamafra: I find it quite helpful and I think it helps me understand the motivating need much better.